armory/experiment.py

```diff
@@ -0,0 +1,155 @@
+from pydantic import BaseModel
+from armory.logs import log
+import os
+import yaml
+import json
+from armory.utils import parse_overrides
+from importlib import import_module
+
+# TODO: Update this so that validation occurs at lowest
+#  level possible.  I.e. we do NOT validate at edge,
+#  and let process throw exception, therefore data model
+#  is assumed to be "unsafe" and each function needs to
+#  protect its inputs.
+
+
+class AttackParameters(BaseModel):
+    """Armory Data Class for `Attack` Parameters
+
+    """
+
+    name: str
+    module: str
+    knowledge: str
+    kwargs: dict
+    type: str = None
+
+
+class DatasetParameters(BaseModel):
+    """Armory Dataclass For `Dataset` Parameters"""
+
+    name: str
+    module: str
+    framework: str
+    batch_size: int
+
+
+class DefenseParameters(BaseModel):
+    """Armory Dataclass for `Defense` Parameters"""
+
+    name: str
+    type: str
+    module: str
+    kwargs: dict
+
+
+class MetricParameters(BaseModel):
+    """Armory Dataclass for Evaluation `Metric` Parameters"""
+
+    means: bool
+    perturbation: str
+    record_metric_per_sample: bool
+    task: list
+
+
+class ModelParameters(BaseModel):
+    """Armory Dataclass for `Model` Parameters"""
+
+    name: str
+    module: str
+    weights_file: str = None
+    wrapper_kwargs: dict
+    model_kwargs: dict
+    fit_kwargs: dict
+    fit: bool
+
+
+class ScenarioParameters(BaseModel):
+    """Armory Dataclass for `Scenario` Parameters"""
+
+    function_name: str
+    module_name: str
+    kwargs: dict
+
+
+class SystemConfigurationParameters(BaseModel):
+    """Armory Dataclass for Environment Configuration Paramters"""
+
+    docker_image: str = None
+    gpus: str = None
+    external_github_repo: str = None
+    output_dir: str = None
+    output_filename: str = None
+    use_gpu: bool = False
+
+
+class MetaData(BaseModel):
+    name: str
+    author: str
+    description: str
+
+
+class PoisonParameters(BaseModel):
+    pass
+
+
+class ExperimentParameters(BaseModel):
+    """Armory Dataclass for Experiment Parameters"""
+
+    _meta: MetaData
+    poison: PoisonParameters = None
+    attack: AttackParameters = None
+    dataset: DatasetParameters
+    defense: DefenseParameters = None
+    metric: MetricParameters = None
+    model: ModelParameters
+    scenario: ScenarioParameters
+    # sysconfig: SystemConfigurationParameters = None
+
+    # def save(self, filename):
+    #     with open(filename, "w") as f:
+    #         f.write(self.json())
+
+    @classmethod
+    def load(cls, filename, overrides=[]):
+        overrides = parse_overrides(overrides)
+        valid_ext = (".aexp", ".json")
+        fname, fext = os.path.splitext(filename)
+        log.info(f"Attempting to Load Experiment from file: {filename}")
+        if fext == ".json":
```

Comment:
Confusing polymorphism. Don't give the user two options when one will do, especially since you can yaml.load scenario_configs/cifar10_baseline.json because yaml is a superset of json.